service/src/main/java/bio/terra/tanagra/service/artifact/ConceptSetService.java

```diff
@@ -103,4 +105,38 @@ public ConceptSet updateConceptSet(
         conceptSetId, userEmail, displayName, description, criteria, outputAttributesPerEntity);
     return conceptSetDao.getConceptSet(conceptSetId);
   }
+
+  public ConceptSet cloneConceptSet(
+      String studyId,
+      String conceptSetId,
+      String userEmail,
+      String displayName,
+      String description) {
+    ConceptSet original = getConceptSet(studyId, conceptSetId);
+    ConceptSet.Builder conceptSetBuilder =
+        ConceptSet.builder()
+            .underlay(original.getUnderlay())
+            .displayName(displayName)
+            .description(description)
+            .createdBy(userEmail)
+            .lastModifiedBy(userEmail);
+
+    List<Criteria> clonedCriteria =
+        original.getCriteria().stream().map(Criteria::deepCopy).toList();
+    Map<String, List<String>> clonedAttrsMap =
+        original.getExcludeOutputAttributesPerEntity().entrySet().stream()
+            .map(
+                attrEntry ->
+                    new AbstractMap.SimpleEntry<>(
+                        attrEntry.getKey(), new ArrayList<>(attrEntry.getValue())))
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
Are these copies necessary? They're not changing anything about the objects (e.g. the id) before writing them to the DB.

Reply:
Criteria has an id, so best to make a deep copy - this is a public function and simply from a design perspective it might not be a bad idea to make a deep copy of all fields because we dont necessarily know who the caller is, or what the caller does. This is the general approach used in workbench services (may be for additional of other reasons as well). hence unless there are blocking concerns with a deep copy, i would prefer to do so. Lmk if you prefer it changed

Reply:
Updated as per discussion and added a comment